File: 0114-Flatten-Binary-Tree-to-Linked-List/Python/Solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

"""
https://leetcode.cn/problems/flatten-binary-tree-to-linked-list/
"""
class Solution:
    def __init__(self):
        self.dummyNode = TreeNode()
        self.curr = self.dummyNode


    def flatten(self, root: TreeNode) -> None:
        self.preorder(root)


    def preorder(self, root: TreeNode) -> None:
        if not root:
            return
        
        left = root.left
        right = root.right
        self.curr.right = root
        self.curr = root
        root.left = None
        
        self.preorder(left)
        self.preorder(right)
        

File: 0114-Flatten-Binary-Tree-to-Linked-List/Java/Solution.java

public class Solution {

    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    TreeNode dummyNode = new TreeNode(-1);
    TreeNode curr = dummyNode;

    /**
     * https://leetcode.cn/problems/flatten-binary-tree-to-linked-list/
     * @param root
     */
    public void flatten(TreeNode root) {
        preorder(root);
    }

    private void preorder(TreeNode root) {
        if (root == null) {
            return;
        }
        TreeNode left = root.left;
        TreeNode right = root.right;
        curr.right = root;
        curr = root;
        root.left = null;

        preorder(left);
        preorder(right);
    }

}